app/src/main/java/com/xmartlabs/template/ui/common/BaseRecyclerViewAdapter.java

```diff
@@ -0,0 +1,204 @@
+package com.xmartlabs.template.ui.common;
+
+import android.support.annotation.CallSuper;
+import android.support.annotation.LayoutRes;
+import android.support.annotation.MainThread;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.v7.widget.RecyclerView;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import com.xmartlabs.template.helper.CollectionHelper;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+/**
+ * A Base Recycler View Adapter with already implemented functions such as
+ * Setting, removing, adding, getting items count.
+ */
+@AllArgsConstructor
+@NoArgsConstructor
+public abstract class BaseRecyclerViewAdapter<T, VH extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<VH> {
+  protected static final int ITEM_TYPE_DIVIDER = -1;
+  protected static final int ITEM_TYPE_ITEM = -2;
+
+  @Getter(AccessLevel.PROTECTED)
+  @NonNull
+  private List<? extends T> items = new ArrayList<>();
+
+  /**
+   * Used to inflate the view layout/elements.
+   *
+   * @param parent      the parent viewgroup
+   * @param layoutResId the layout resource id
+   * @return the inflated view
+   */
+  protected View inflateView(@NonNull ViewGroup parent, @LayoutRes int layoutResId) {
+    return LayoutInflater.from(parent.getContext()).inflate(layoutResId, parent, false);
+  }
+
+  /**
+   * Used to set the items data for the recycler view and notifying any registered observers that the data set has
+   * changed.
+   *
+   * @param items the items that will be the data for the recycler view
+   */
+  @MainThread
+  public void setItems(@Nullable List<? extends T> items) {
+    this.items = items == null ? new ArrayList<>() : items;
+    notifyDataSetChanged();
+  }
+
+  /**
+   * Used to set the items data for the recycler view in form of an array and calls {@link #setItems(List)}
+   *
+   * @param items array of items that will be the data for the recycler view
+   */
+  @MainThread
+  public void setItems(@NonNull T[] items) {
+    setItems(Arrays.asList(items));
+  }
+
+  /**
+   * Used to remove an item from the data and any registered observers of its removal
+   *
+   * @param item the item to be removed
+   */
+  @MainThread
+  public void removeItem(@NonNull T item) {
+    int position = items.indexOf(item);
+    if (position >= 0) {
+      items.remove(position);
+      notifyDataSetChanged();
+    }
+  }
+
+  /**
+   * Used to add items to the data for the recycler view and notifies any registered observers that the data set has
+   * changed.
+   *
+   * @param items the items that will be the data for the recycler view
+   */
+  @MainThread
+  public boolean addItems(@Nullable List<? extends T> items) {
```

Comment:
I think maybe keeping `@Nullable` is fine here. Serves the purpose of not having to check Null when calling it